### PR TITLE
Change ObjectStoreRegistry from struct to trait to provide polymorphism

### LINKED ISSUE #5541: Introduce ObjectStoreManager trait for the ObjectStoreRegistry to provide polymorphism for get_by_url
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The context is we are working on introducing a data source cache layer for Ballista & DataFusion https://github.com/apache/arrow-ballista/issues/645. To make the cache layer transparent for the `ObjectStore` caller, we plan to introduce a special cached based `ObjectStore`, which will search local cache first and then the original remote source `ObjectStore`. A sample implementation is like https://github.com/yahoNanJing/arrow-ballista/blob/dev-20230302/ballista/core/src/cache_layer/object_store/file.rs. Then we hope the `ObjectStoreRegistry` be able to provide polymorphism for the `get_by_url` method. If cached enabled, that function will always return a cached-based `ObjectStore`.

**Describe the solution you'd like**

To introduce an object trait as the attribute of `ObjectStoreRegistry`.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by yahoNanJing**
# Which issue does this PR close?

Closes #5541.

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by yahoNanJing** [comment]
Hi @alamb, @tustvold, @thinkharderdev and @andygrove, an initial prototype of the data source cache layer for Ballista has almost been implemented https://github.com/yahoNanJing/arrow-ballista/tree/dev-20230302. From now on, I will continuously reorganize the code and contribute back to the community. 

To achieve this, we need to apply a few changes to the datafusion and arrow-rs. Could you help review this PR, first?

**Review by tustvold — COMMENTED** [review]
This seems to have non-trivial overlap with ObjectStoreProvider, I wonder if we could combine them?

On a somewhat related note, I'm a little confused why this is needed for a cache system, couldn't that just be kept as a detail of whatever is creating the ObjectStore? I.e. the stores are registered already wrapped in a CachingStore (or whatever it gets called). I think this would be significantly simpler and easier to understand, and would not require any changes outside of Ballista?

**Comment by yahoNanJing** [comment]
> This seems to have non-trivial overlap with ObjectStoreProvider, I wonder if we could combine them?

The `ObjectStoreProvider` and `ObjectStoreManager` are for different aspects. As described in the code, `ObjectStoreProvider` focuses to get `ObjectStore` when it's not found in cache. While `ObjectStoreManager` focuses on the polymorphism of how to get an `ObjectStore` by `get_by_url`. The main reason to introduce `ObjectStoreManager` here, is not to break existing usage of `ObjectStoreProvider`. However, if not considering of existing usage of `ObjectStoreProvider`, I have no preference on whether to combine them into one.

File: datafusion/execution/src/object_store.rs
Comment (thinkharderdev) [inline datafusion/execution/src/object_store.rs:173]: I'm confused by this. This is the same interface as `ObjectStoreRegistry`. What is the functional difference that requires wrapping this in another layer of indirection?

**Comment by yahoNanJing** [comment]
An example usage is linked on the issue description. https://github.com/yahoNanJing/arrow-ballista/blob/dev-20230302/ballista/core/src/cache_layer/object_store/file.rs

**Comment by tustvold** [comment]
> An example usage is linked on the issue description. https://github.com/yahoNanJing/arrow-ballista/blob/dev-20230302/ballista/core/src/cache_layer/object_store/file.rs

Why can't whatever is creating the inner `ObjectStore` that is passed to https://github.com/yahoNanJing/arrow-ballista/blob/dev-20230302/ballista/core/src/cache_layer/object_store/file.rs#L45 just also create the `FileCacheObjectStore`?

File: datafusion/execution/src/object_store.rs
Comment (yahoNanJing) [reply datafusion/execution/src/object_store.rs:173]: The reason for this is to make the cache transparent for the caller. An example usage is here,
https://github.com/yahoNanJing/arrow-ballista/commit/48c703b3c62235d5e70f19a1bb7a39d9584138ee

Comment (yahoNanJing) [reply datafusion/execution/src/object_store.rs:173]: In Ballista, we can create an `ObjectStoreRegistry` with a `CacheBasedObjectStoreManager`. Then every `ObjectStore` got by `ObjectStoreRegistry` will be wrapped with a cache layer.

**Comment by yahoNanJing** [comment]
> Why can't whatever is creating the inner ObjectStore that is passed to https://github.com/yahoNanJing/arrow-ballista/blob/dev-20230302/ballista/core/src/cache_layer/object_store/file.rs#L45 just also create the FileCacheObjectStore and register the FileCacheObjectStore directly? This is the same story as for other ObjectStore decorators that add metrics, tracing, etc...

The reason is to make the cache layer transparent for the caller. To achieve this, when invoking `get_by_url`, we need to get a cache layer wrapped `ObjectStore`

**Comment by tustvold** [comment]
> The reason is to make the cache layer transparent for the caller.

Who is the caller in this case? I had understood Ballista to have a config system that handled object store registration, in order to allow it to propagate to the workers correctly?

> when invoking get_by_url, we need to get a cache layer wrapped ObjectStore

This is already covered by ObjectStoreProvider, no? You can just have an ObjectStoreProvider that returns decorated ObjectStore?

**Comment by yahoNanJing** [comment]
> This is already covered by ObjectStoreProvider, no? You can just have an ObjectStoreProvider that returns decorated ObjectStore?

When you have already cached the `ObjectStore` or register the `ObjectStore` manually, the `ObjectStoreProvider` will not be effective. Then there's no transparency.

File: datafusion/execution/src/object_store.rs
Comment (alamb) [inline datafusion/execution/src/object_store.rs:107]: I am sorry I am missing something -- how is this different than 

https://github.com/apache/arrow-datafusion/blob/860918d17b6bde396b04d718ee1c76d93054bf11/datafusion/execution/src/object_store.rs#L89

?

> While ObjectStoreManager focuses on the polymorphism of how to get an ObjectStore by get_by_url.

It seems like the only function of `ObjectStoreProvider` is a polymorphic `get_by_url` 🤔

**Comment by yahoNanJing** [comment]
> Is that not an advantage, for example, if registering a InMemory or LocalFileSystem, I would not want them to be automatically wrapped in a caching decorator?

I don't think it's an advantage. How to register an `ObjectStore` should also be polymorphic. For example, by introducing the `ObjectStoreManager`, we will be able to use `InMemory` as the cache layer for `LocalFileSystem` instead of using `LocalFileSystem` directly.

**Comment by yahoNanJing** [comment]
> Does Ballista have some mechanism to manually register stores from config / SQL or something?

It depends. For Ballista client, it can manually register stores. And we should not have the assumption that users don't do the registration before using the auto-configured stores registered by `ObjectStoreProvider`.

File: datafusion/execution/src/object_store.rs
Comment (yahoNanJing) [reply datafusion/execution/src/object_store.rs:107]: Thanks @alamb for reviewing. Actually, the main concern is that `ObjectStoreProvider` only covers the `get_by_url`. It does not cover the manual registration. Therefore, in this PR, I introduced `ObjectStoreManager` to cover both. In regard to whether to combine `ObjectStoreManager` and `ObjectStoreProvider`, I have no preference. Here, the main reason of not combining them is for the backward compatibility of existing usage of `ObjectStoreProvider`.

Comment (thinkharderdev) [reply datafusion/execution/src/object_store.rs:107]: `register_store` already exists on `ObjectStoreRegistry` so this would essentially make `ObjectStoreRegistry` and `ObjectStoreProvider` have the same interface which I think is confusing. So as I understand it, the role of `ObjectStoreProvider` is just to allow lazy construction. Manual registration can already be done through the registry. So it just seems to me like we could serve the same use case by holding an `Arc<dyn ObjectStoreRegistry>` in whatever is managing the cache layer.

Comment (alamb) [reply datafusion/execution/src/object_store.rs:107]: I tried to improve the docs in https://github.com/apache/arrow-datafusion/pull/5577

Comment (yahoNanJing) [reply datafusion/execution/src/object_store.rs:107]: > So it just seems to me like we could serve the same use case by holding an Arc<dyn ObjectStoreRegistry> in whatever is managing the cache layer.

Thanks @thinkharderdev. Since now the `ObjectStoreRegistry` is a struct, I introduced another trait `ObjectStoreManager` for this.

**Comment by yahoNanJing** [comment]
Hi @thinkharderdev, @alamb, according to previous discussion, I remove the trait `ObjectStoreProvider`. Instead, I change the `ObjectStoreRegistry` from struct to trait, which will be much more simple and straight forward.

File: datafusion/execution/src/object_store.rs
Comment (alamb) [inline datafusion/execution/src/object_store.rs:82]: We seem to have lost a significant amount of documentation -- maybe it was lost in a merge?

It would be great to restore the docstrings unless there is some reason to remove them

**Review by thinkharderdev — APPROVED** [review]
Nice! I always thought this was a bit convoluted. I think this makes the API more straightforward.

File: datafusion-cli/src/object_storage.rs
Comment (tustvold) [inline datafusion-cli/src/object_storage.rs:86]:
```suggestion
            let store = match ObjectStoreScheme::from_str(url.scheme())?
                    ObjectStoreScheme::S3 => build_s3_object_store(url)?,
                    ObjectStoreScheme::GCS => build_gcs_object_store(url)?,
            };
```

Or something similar

File: datafusion/execution/src/object_store.rs
Comment (tustvold) [inline datafusion/execution/src/object_store.rs:95]: Do we need both a `put_with_url` and `register_store`?

The motivation for not passing the entire URL, was to avoid confusion over what bits of the URL "counted", e.g. should credential influence this? I don't feel strongly, but feel we probably shouldn't have two methods that do basically the same thing

**Review by tustvold — COMMENTED** [review]
Like the switch to a trait object, more flexible and significantly easier to follow.

Only question concerns the overlap of `register_store` and `put_with_url`. The original design intentionally did not allow registering arbitrary URLs to make it clearer what bits of a URL identified an object store.

I don't feel especially strongly about this, but feel we probably shouldn't have overlapping methods, and definitely should document that things like the path should not influence the returned ObjectStore (as this would have very funky semantics)

File: datafusion/execution/src/object_store.rs
Comment (yahoNanJing) [reply datafusion/execution/src/object_store.rs:82]: My bad. I'll raise another commit to back the comments with examples.

**Comment by alamb** [comment]
> Thanks @tustvold for your suggestion. put_with_url and register_store are using different strategies for registration. Perhaps it's better to just use key and value as the parameters for the interface. I'll refine this in the next commit.

@yahoNanJing  do you plan to merge this PR and refine in a follow on one, or shall we hold off merging this PR until you have made more changes?

**Comment by yahoNanJing** [comment]
Hi @alamb, sorry for the delay. I'm going to add commit to this PR for the review comments today

**Comment by yahoNanJing** [comment]
Hi @alamb, @thinkharderdev, @tustvold, just add a commit for your comments. Could you help have a review again?

File: datafusion-cli/src/object_storage.rs
Comment (tustvold) [inline datafusion-cli/src/object_storage.rs:63]: What is the use-case for registering stores not based on scheme and host? I mainly ask as it is unclear how this method interacts with get_by_url? Could we just keep the host and scheme signature instead?

Comment (tustvold) [reply datafusion-cli/src/object_storage.rs:63]: > the key for registration is

That's the key for lookup, and so it was harmless for it to contain more, it would just not be found.

> port

Including an optional port seems like a sensible addition

Comment (yahoNanJing) [reply datafusion-cli/src/object_storage.rs:63]: If with only scheme and host for register, how can we achieve register by url which may be invoked in `get_with_url` for lazy registration with key of `&url[url::Position::BeforeScheme..url::Position::BeforePath]`

Comment (tustvold) [reply datafusion-cli/src/object_storage.rs:63]: That would be a bug then, yes. We shouldn't be registering with that, well spotted.

It should instead call register_store with just the relevant parts of the URL

File: datafusion/execution/src/object_store.rs
Comment (tustvold) [inline datafusion/execution/src/object_store.rs:220]: As pointed out in https://github.com/apache/arrow-datafusion/pull/5543/files#r1143108925 this has a bug, in that it registers using the entire URL authority, including things like credentials, etc... This should instead be calling `register_store` as "normal".

File: datafusion-cli/src/object_storage.rs
Comment (tustvold) [reply datafusion-cli/src/object_storage.rs:63]: Alternatively we could just make the register_store take a `Url` as the key, and leave it to the `ObjectStoreRegistry` to define what parts it uses. I don't know, this API has morphed so much from its original design now :sweat_smile:

Comment (yahoNanJing) [reply datafusion-cli/src/object_storage.rs:63]: I'm OK with either with url or a pair of (scheme, host, port).

Comment (tustvold) [reply datafusion-cli/src/object_storage.rs:63]: Let's go with a URL, it is more flexible, avoids a potentially breaking change even if it was unintentional, and also matches the get method

Comment (tustvold) [reply datafusion-cli/src/object_storage.rs:63]: Thank you, apologies for the back and forth :smile:

Comment (yahoNanJing) [reply datafusion-cli/src/object_storage.rs]: Np. We are making things better and better. Thanks for your suggestions to let the interface simply and meaningful.

File: datafusion/execution/src/object_store.rs
Comment (tustvold) [inline datafusion/execution/src/object_store.rs:118]:
```suggestion
/// lazily by providing a custom implementation of [`ObjectStoreRegistry`]
```

File: datafusion/execution/src/object_store.rs
Comment (tustvold) [inline datafusion/execution/src/object_store.rs:173]:
```suggestion
        object_stores.insert("file://".to_string(), Arc::new(LocalFileSystem::new()));
```

See below comment about removing trailing `/`

File: datafusion/execution/src/object_store.rs
Comment (tustvold) [inline datafusion/execution/src/object_store.rs:141]:
```suggestion
///
/// Stores are registered based on the scheme, host and port of the provided URL
/// with a [`LocalFileSystem::new`] automatically registered for `file://`
///
/// For example:
///
/// - `file:///my_path` will return the default LocalFS store
/// - `s3://bucket/path` will return a store registered with `s3://bucket` if any
/// - `hdfs://host:port/path` will return a store registered with `hdfs://host:port` if any
pub struct DefaultObjectStoreRegistry {
```

File: datafusion/execution/src/object_store.rs
Comment (tustvold) [inline datafusion/execution/src/object_store.rs:190]:
```suggestion
```

These docs aren't rendered by docs.rs, I moved this onto the type definition and reworded it slightly

File: datafusion-cli/src/object_storage.rs
Comment (tustvold) [inline datafusion-cli/src/object_storage.rs:49]:
```suggestion
/// An [`ObjectStoreRegistry`] that can automatically create S3 and GCS stores for a given URL
#[derive(Debug, Default)]
```

**Review by tustvold — APPROVED** [review]
Mostly just some minor nits, looks good to me, thank you

**Comment by alamb** [comment]
Thanks for sticking with this @yahoNanJing and @tustvold

File: datafusion-cli/src/object_storage.rs
Comment (yahoNanJing) [reply datafusion-cli/src/object_storage.rs:69]: Agree.

File: datafusion/execution/src/object_store.rs
Comment (yahoNanJing) [reply datafusion/execution/src/object_store.rs:216]: My bad. The trailing `/` can be removed.

**Pull request merged**
